app/models/upload_snapshot.rb

```diff
@@ -6,9 +6,15 @@ class UploadSnapshot < ApplicationRecord
   alias_attribute :existing_files, :files
 
   def snapshot_deletions(work_changes, s3_filenames)
+    s3_filenames_sorted = s3_filenames.sort
     existing_files.each do |file|
       filename = file["filename"]
-      unless s3_filenames.include?(filename)
+      # Use Ruby's Binary Search functionality instead of a plain Ruby Array `.include?`
+      # to detect missing values in the array because the binary search performs
+      # much faster when the list of files is large. Notice that the binary search
+      # requires that the list of files is sorted.
+      # See https://ruby-doc.org/3.3.6/bsearch_rdoc.html
+      if s3_filenames_sorted.bsearch { |s3_filename| filename <=> s3_filename }.nil?
```

Comment:
This is the other optimization, notice the use of Binary Search `bsearch` instead of a normal `include?`